report.md

```diff
@@ -1,7 +1,60 @@
 # Subsystem Design Report
 
 ## Subsystem 1 (MFA)
-
+
+MFA Tag Interfaces:
+1. RFDuino BLE API  
+The RFduino device must include the #include <RFduinoBLE.h> command in order to make use of the RFduino BLE. 
+1. Arduino Cryptography Libraries
+
+Web App Interface:
+1. Web Bluetooth API  
+Access to Web Bluetooth API will be subject to permissions and it will only work in secure contexts. It only works with sites on HTTPS.  The connection to the device can only be called by a  user action like a click for example. 
+
+Communication protocol between the two devices:  
+Digital Signature + Diffie Hellman Key Exchange + Symmetric Encryption
+
+Algorithms to be used:
+1. Ed25519 for digital signature - to be used for the tag from [Arduino Cryptography Libraries](https://rweather.github.io/arduinolibs/crypto.html) and for web app from [NaCL](https://github.com/dchest/tweetnacl-js#public-key-authenticated-encryption-box) and [Libsodium](https://github.com/jedisct1/libsodium.js)
+1. Curve25519 for getting shared key - to be used for web app from [here](https://github.com/indutny/elliptic)
+1. AES256 for encryption 
+1. SHA256 for hashing
+
+Rationale:  
+A digital signature ensures integrity, authentication and non-repudiation. We cannot afford to use RSA on the rfduino board due to the lack of RAM. In fact, ed25519 seems to have many [benefits](https://risan.io/upgrade-ssh-key-to-ed25519.html). 
+
+
+##### Assumptions to be made before any communication is made:
+Private and public keys that are stored initially are generated via the ED25519.
+
+Before the communication begins:  
+The database contains:
+1. A hash of the private key with salt linked to the user
+
+The MFA tag contains:
+1. A private key linked to the user
+1. The web app’s public key
+
+The web app contains: 
+1. Its own private key 
+1. Its own public key 
+
+
+For 2FA Login into Web App: (only MFA tag needs to be authenticated to belong to the user)
+1. User logs in with the correct log in credentials - username and password. 
+1. After that, the web app will request to pair with a BLE device nearby - which is the user’s 2FA token.  
+1. To authenticate that the MFA tag really belongs to the user logging in, we need to ensure that the MFA tag has the private key of the user. The MFA tag will use its private key to sign a hash of a random message. 
+1.  The web app and the MFA tag will then get a secret shared key using Diffie Helman Key Exchange. 
+1. The MFA tag would then use AES with the shared key to encrypt the digital signature and send it over. 
+1. The web app would then decrypt and use the user’s public key to attain the hash value of the random message. It would also compute the hash of that random message and then it would then see if the hashes match. If they match, it means that the MFA tag has the private key in it and it belongs to the user because only it is able to sign it with the private key. 
+1. Thus, the web server can log the user in. 
+
```

Comment:
Just to confirm the steps.

1. Message generated by web app and sent to BLE device
2. BLE device "on receive" will encrypt the message with private key stored in it
3. Computer will decrypt message using the public key
4. generate hashes of both messages then compare?

Reply:
2. BLE device "on receive" will sign the message with its private key then encrypt it using the shared key that they already decided on. 
3. Computer will decrypt using the shared key then check if the hash of the message is correct by using the public key to decrypt the digital signature.